### PR TITLE
React frontend proposal

### PR DESCRIPTION
**Pull request by sujankapadia**
Summary

Complete migration from the legacy Streamlit dashboard to a modern React + FastAPI architecture. The app is now a single installable artifact that serves both the API  and React SPA on port 8000.

  Frontend

  - React SPA with Vite, TypeScript, TanStack Query, Tailwind CSS v4, and virtual scrolling
  - Pages: Dashboard, Sessions, Session Detail, Search, Analytics, Analysis, Find Examples, Import, Active Sessions, Bookmarks
  - Command Palette (Cmd+K) with fuzzy search across pages, projects, sessions, and FTS content
  - Real-time updates via SSE — file watcher auto-imports new sessions and pushes cache invalidation to the UI

  Backend

  - FastAPI REST API with 22 GET and 4 POST endpoints across 8 routers
  - File watcher auto-imports new sessions from ~/.claude/projects/ with catch-up on startup
  - SPA routing — catch-all fallback serves index.html for client-side routes in production

  Key Features Added

  - Find Examples — natural language discovery of prompts and sessions (hybrid FTS + LLM ranking)
  - Analysis Scoping — analyze entire sessions, time ranges, or search hit context windows
  - Generic OpenAI-compatible provider — supports OpenRouter, Ollama, and any OpenAI-compatible API
  - Search scope tabs — always-visible All/Messages/Tool Input/Tool Results tabs
  - Conversation role filter — filter by All/User/Assistant in the viewer

  Architecture Cleanup

  - Removed Streamlit entirely — dashboard UI, streamlit/pandas/altair dependencies, run_dashboard.sh, claude-code-streamlit entry point
  - Restructured package — moved streamlit_app/services/ → claude_code_analytics/services/, streamlit_app/models/ → claude_code_analytics/models/
  - Consolidated entry points — claude-code-analytics now launches FastAPI + React (port 8000)
  - Updated install/uninstall scripts — install checks Node.js 18+, builds frontend; uninstall cleans dist/ and node_modules/
  - Fixed 5 pre-existing test failures — all 161 tests pass

  Documentation

  - README rewritten for React + FastAPI stack (removed all Streamlit references)
  - CHANGELOG updated with all additions, changes, and removals
  - CLAUDE.md updated with new architecture
  - TESTING.md rewritten
  - SECURITY.md updated

  Test Plan

  - All 161 tests pass (pytest tests/ -q)
  - Install/uninstall tested with isolated HOME and venv
  - Frontend builds successfully (npm run build)
  - SPA routing works in production mode

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Launched React single-page application with real-time updates via SSE, dark mode support, and improved UX (virtual scrolling, command palette, minimap navigation)
  * Added session bookmarking within conversations
  * Added hybrid session similarity search combining full-text and semantic search
  * Added active session monitor showing currently running Claude instances
  * Enhanced in-conversation search with match highlighting and keyboard navigation

* **Changed**
  * Migrated from Streamlit dashboard to FastAPI backend serving React frontend on port 8000
  * Increased Python minimum version requirement from 3.9 to 3.10

* **Removed**
  * Streamlit dashboard and related dependencies

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -76,22 +103,31 @@ if [ ! -f "$CONFIG_FILE" ]; then
...
     chmod 600 "$CONFIG_FILE"
 fi
 
-# Install Python package
+# ── Install Python package ───────────────────────────────────────────
+
 echo "📦 Installing Python package and dependencies..."
 if command -v pip3 &> /dev/null; then
     pip3 install -e "$SCRIPT_DIR"
     echo -e "${GREEN}✓ Installed claude-code-analytics package${NC}"
```
[the remark this answers is not in the document]
Comment (skapadia) [reply install.sh]: **Dismissed** — Test style suggestion, not a bug.

File: claude_code_analytics/api/routers/examples.py [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+    fts_query = " ".join(terms)
+    try:
+        if req.role or req.scope == "Messages":
+            # Use message-specific search with optional role filter
+            fts_results_raw = db.search_messages(
+                query=fts_query,
+                project_id=req.project_id,
+                role=req.role,
+                limit=200,
+            )
```
[the remark this answers is not in the document]
Comment (skapadia) [reply claude_code_analytics/api/routers/examples.py]: **Dismissed** — This is in `find_example_sessions`, not the search router. The role filter is intentional here — examples always search messages to find shareable user prompts.

File: claude_code_analytics/api/services/import_service.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+                                        "message_index": current_index,
+                                    }
+                            elif item.get("type") == "tool_result":
+                                tool_id = item.get("tool_use_id")
+                                if tool_id:
+                                    tool_results[tool_id] = {
+                                        "content": item.get("content"),
+                                        "is_error": item.get("is_error", False),
+                                    }
+
```
[the remark this answers is not in the document]
Comment (skapadia) [reply claude_code_analytics/api/services/import_service.py]: **Deferred** as #28 — edge case with low impact.

File: tests/test_import_service.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        try:
+            config_mod.CLAUDE_CODE_PROJECTS_DIR = tmp_path / "projects"
+            config_mod.DATABASE_PATH = Path(db_with_schema)
+
+            loop = asyncio.new_event_loop()
+            try:
+                result = loop.run_until_complete(run_import(event_bus, db_path=db_with_schema))
+                # Should complete without SQLite threading errors
+                assert result["sessions"] >= 0
+            finally:
```
[the remark this answers is not in the document]
Comment (skapadia) [reply tests/test_import_service.py]: **Dismissed** — Test style suggestion, not a bug.

File: tests/test_app.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        from claude_code_analytics.api.app import create_app
+
+        app = create_app()
+
+        @asynccontextmanager
+        async def _noop_lifespan(a: FastAPI):
+            yield
+
+        app.router.lifespan_context = _noop_lifespan
+        yield TestClient(app)
```
[the remark this answers is not in the document]
Comment (skapadia) [reply tests/test_app.py]: **Dismissed** — Test style suggestion, not a bug.

File: claude_code_analytics/api/routers/examples.py [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+        else:
+            fts_results = db.search_all(
+                query=fts_query,
+                project_id=req.project_id,
+                limit=200,
+            )
+        for r in fts_results:
+            candidate_ids.add(r["session_id"])
+    except Exception:
+        pass  # FTS may fail on unusual queries; continue with other signals
```
[the remark this answers is not in the document]
Comment (skapadia) [reply claude_code_analytics/api/routers/examples.py]: **Dismissed** — The `pass` is intentional. FTS can fail on unusual query syntax, and we fall through to other search strategies (tool pattern matching). Surfacing the error would break valid queries.

File: claude_code_analytics/api/services/process_scanner.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    for proc in psutil.process_iter(["pid", "name", "cmdline", "cwd", "create_time", "terminal"]):
+        try:
+            info = proc.info
+            name = info.get("name", "")
+            cmdline = info.get("cmdline") or []
+
+            # Match the main claude process
+            # The process name is "claude" or the first cmdline arg ends with "claude"
+            is_claude = name == "claude" or (cmdline and Path(cmdline[0]).name == "claude")
+            if not is_claude:
```
[the remark this answers is not in the document]
Comment (skapadia) [reply claude_code_analytics/api/services/process_scanner.py]: **Dismissed** — This is a local-only macOS/Linux developer tool. No Windows support planned.

File: frontend/src/pages/session-detail.tsx [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+  const {
+    data: session,
+    isPending: sessionLoading,
+    error: sessionError,
+  } = useQuery({
+    queryKey: ["session", id],
+    queryFn: () => fetchSession(id!),
+    enabled: !!id,
+  });
```
[the remark this answers is not in the document]
Comment (skapadia) [reply frontend/src/pages/session-detail.tsx]: **Dismissed** — TanStack Query's `enabled: !!id` already prevents the fetch when `id` is undefined. The route pattern guarantees `id` is present.
